### PR TITLE
8292044: HttpClient doesn't handle 102 or 103 properly

### PR DESCRIPTION
**Pull request by jaikiran**
Can I please get a review of this change which proposes to fix https://bugs.openjdk.org/browse/JDK-8292044?

The linked JBS issue notes two parts to fixing this. Part one is to (internally) ignore the intermediate 1xx informational responses, in the client and wait for subsequent final response from the server. Part two is to introduce newer APIs to let applications using HttpClient, to have access to these intermediate response (codes). This commit (only) addresses part one. Part two is out of scope of this change and a separate issue will be opened to address it (at a later time).

The commit in this PR introduces a check to see if the returned response is an informational response (as defined by RFC-2616 https://www.rfc-editor.org/rfc/rfc2616#page-58). If the response code is between 102 and 199 (inclusive), then this change ignores that response and keeps waiting for a subsequent final response from the server.

The request timeout (if set) will _not_ be reset when a intermediate informational response is received (and we ignore it). The request timeout handling continues to be the same as what it is currently and will span from the request start till the final response is received. If no final response is received within the duration of request timeout (if set) then the application will continue to receive a request timeout exception.

A new test class has been introduced to reproduce the issue and test the fix. The test tests both HTTP/1.1 and HTTP2. 

tier1, tier2 and tier3 testing is in progress.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8292044](https://bugs.openjdk.org/browse/JDK-8292044): HttpClient doesn't handle 102 or 103 properly

### Reviewers
 * @reschke (no known github.com user name / role) ⚠️ Review applies to [394ac932](https://git.openjdk.org/jdk/pull/10169/files/394ac932fc72bf0402d8f75066ce3090b2a32d71)
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**)
 * [Chris Hegarty](https://openjdk.org/census#chegar) (@ChrisHegarty - **Reviewer**) ⚠️ Review applies to [fda47d20](https://git.openjdk.org/jdk/pull/10169/files/fda47d2000f8e4efe42a44da3e6f8e0e9660fa5a)
 * [Michael McMahon](https://openjdk.org/census#michaelm) (@Michael-Mc-Mahon - **Reviewer**) ⚠️ Review applies to [fda47d20](https://git.openjdk.org/jdk/pull/10169/files/fda47d2000f8e4efe42a44da3e6f8e0e9660fa5a)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10169/head:pull/10169` \
`$ git checkout pull/10169`

Update a local copy of the PR: \
`$ git checkout pull/10169` \
`$ git pull https://git.openjdk.org/jdk pull/10169/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10169`

View PR using the GUI difftool: \
`$ git pr show -t 10169`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10169.diff">https://git.openjdk.org/jdk/pull/10169.diff</a>

</details>

### CODE REVIEW

File: src/java.net.http/share/classes/jdk/internal/net/http/Exchange.java [hunk 1]
```diff
@@ -472,10 +472,56 @@ private CompletableFuture<Response> sendRequestBody(ExchangeImpl<T> ex) {
...
+        // we handle it specifically as defined in the RFC-2616 (HTTP 1.1 spec). Any other
+        // 1xx response code isn't part of the HTTP 1.1 spec and the spec states that
+        // these 1xx response codes are informational and ignored and the clients are
+        // expected to wait for the actual response (headers) to be sent back at a later
+        // time.
+        if (statusCode >= 102 && statusCode <= 199) {
+            if (debug.on()) {
+                debug.log("Ignoring (1xx informational) response code "
+                        + rsp.statusCode());
+            }
```
Comment (dfuch) [inline src/java.net.http/share/classes/jdk/internal/net/http/Exchange.java:511]: We should probably log that using one of the Log.xxx methods too ?
Have the response headers in Response already been logged (using Log) by the time we reach here? If not we need to log them.

Comment (jaikiran) [reply src/java.net.http/share/classes/jdk/internal/net/http/Exchange.java:512]: Hello Daniel,

> We should probably log that using one of the Log.xxx methods too ?

I've now updated the PR to add the `Log.logTrace` call too.

> Have the response headers in Response already been logged (using Log) by the time we reach here?

Yes, both for HTTP1 and HTTP2 the response headers get logged as soon as a `Response` instance is created here https://github.com/openjdk/jdk/blob/master/src/java.net.http/share/classes/jdk/internal/net/http/Http1Response.java#L209 and here https://github.com/openjdk/jdk/blob/master/src/java.net.http/share/classes/jdk/internal/net/http/Stream.java#L516. So by the time, the control reaches here those response headers will already be logged. Do note that the spec doesn't mandate any response headers for these informational responses. So there may not be any response headers.

**Comment by jaikiran** [comment]
The PR has been updated to include a new test to verify that the request timeout isn't impacted when informational responses are involved and the application receives the timeout exception if the server doesn't respond with a final response, within the timeout duration.

tier1, tier2 and tier3 testing passed with the changes in this PR.

**Comment by Michael-Mc-Mahon** [comment]
We probably should consider how a future API change would work, which allows an application to receive the intermediate responses and make sure this change doesn't cause any problems for that. Though, I don't see an issue currently. Regarding a new API, maybe a method could be added to HttpResponse.BodyHandler to supply intermediate responses?

**Comment by jaikiran** [comment]
A new JBS issue (https://bugs.openjdk.org/browse/JDK-8293574) got reported for this same bug. The reporter there provided test cases to reproduce the issue. Those tests helped identify an additional fix that was needed in this PR.

Specifically, the RFC states that if the request is configured for "Expect-Continue" and if the server still sends a `100` response, then the client should ignore it and wait for a final response. This wasn't being done in this proposed patch. I've now updated this PR to handle that case and have also updated the test case to cover this scenario.

**Comment by jaikiran** [comment]
> 
> Specifically, the RFC states that if the request is configured for "Expect-Continue" and if the server still sends a `100` response....

That should read "the RFC states that if the request is *not* configured for "Expect-Continue" ..."

File: src/java.net.http/share/classes/jdk/internal/net/http/Exchange.java [hunk 2]
```diff
@@ -472,10 +472,62 @@ private CompletableFuture<Response> sendRequestBody(ExchangeImpl<T> ex) {
...
+     * </p>
+     *
+     * @param rsp The response
+     * @return A {@code CompletableFuture} with the final response from the server
+     */
+    private CompletableFuture<Response> ignore1xxResponse(final Response rsp) {
+        final int statusCode = rsp.statusCode();
+        // we ignore any response code which is 1xx.
+        // For 100 (with the request configured to expect-continue) and 101, we handle it
+        // specifically as defined in the RFC-2616 (HTTP 1.1 spec), outside of this method.
```
Comment (reschke) [inline src/java.net.http/share/classes/jdk/internal/net/http/Exchange.java:503]: RFC 2616 is irrelevant. You need to look at RFC 9110 and 9112.

File: src/java.net.http/share/classes/jdk/internal/net/http/Exchange.java [hunk 3]
```diff
@@ -472,10 +472,62 @@ private CompletableFuture<Response> sendRequestBody(ExchangeImpl<T> ex) {
...
+        // As noted in RFC-7231, section 6.2.1, if response code is 100 and if the request wasn't
```
Comment (reschke) [inline src/java.net.http/share/classes/jdk/internal/net/http/Exchange.java:504]: So is RFC 7231.

File: src/java.net.http/share/classes/jdk/internal/net/http/Exchange.java [hunk 4]
```diff
@@ -472,10 +472,62 @@ private CompletableFuture<Response> sendRequestBody(ExchangeImpl<T> ex) {
...
+        // configured with expectContinue, then we ignore the 100 response and wait for the final
+        // response (just like any other 1xx response)
+        // Any other response code between 102 and 199 (both inclusive) aren't specified in the
+        // HTTP 1.1 spec. The spec states that these 1xx response codes are informational and
```
Comment (reschke) [inline src/java.net.http/share/classes/jdk/internal/net/http/Exchange.java:508]: This is not specific for version 1.1 of HTTP. See RFC 9110.

**Review by reschke — COMMENTED** [review]
Left a few comments regarding which RFCs are relevant.

**Comment by jaikiran** [comment]
Hello Michael,

> Regarding a new API, maybe a method could be added to HttpResponse.BodyHandler to supply intermediate responses?

I had a look at the HttpClient API and as you note, we could add a new method on `BodyHandler` to provide the application with these interim responses. Something like:

```
/**
 * Invoked when the client receives an interim informational (1xx) response for
 * a request.
 * @param responseInfo The interim response
 */
default void applyInterim(ResponseInfo responseInfo) {
    // do nothing by default
}
```
I plan to experiment with this in a separate JBS issue.

File: src/java.net.http/share/classes/jdk/internal/net/http/Exchange.java (hunk 2 above)
Comment (jaikiran) [reply src/java.net.http/share/classes/jdk/internal/net/http/Exchange.java]: Thank you Julian for those references. I've cleaned up this comment in the updated version of this PR.

File: src/java.net.http/share/classes/jdk/internal/net/http/Exchange.java [hunk 5]
```diff
@@ -472,10 +472,62 @@ private CompletableFuture<Response> sendRequestBody(ExchangeImpl<T> ex) {
...
+        final int statusCode = rsp.statusCode();
+        // we ignore any response code which is 1xx.
+        // For 100 (with the request configured to expect-continue) and 101, we handle it
+        // specifically as defined in the RFC-9110, outside of this method.
+        // As noted in RFC-9110, section 15.2.1, if response code is 100 and if the request wasn't
+        // configured with expectContinue, then we ignore the 100 response and wait for the final
+        // response (just like any other 1xx response)
+        // Any other response code between 102 and 199 (both inclusive) aren't specified in the
+        // "HTTP semantics" RFC-9110. The spec states that these 1xx response codes are informational
+        // and interim and the client will ignore them and will continue to wait to receive the
```
Comment (reschke) [inline src/java.net.http/share/classes/jdk/internal/net/http/Exchange.java:509]: I'd rephrase that as "clients *can' ignore them if they choose to"

**Review by reschke — APPROVED** [review]

File: test/jdk/java/net/httpclient/Response1xxTest.java [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+    private String http2RequestURIBase;
+
+    @BeforeClass
+    public void setup() throws Exception {
+        serverSocket = new ServerSocket(0, 0, InetAddress.getLoopbackAddress());
+        server = new Http11Server(serverSocket);
+        new Thread(server).start();
+        http1RequestURIBase = "http://" + serverSocket.getInetAddress().getHostAddress()
+                + ":" + serverSocket.getLocalPort();
+
```
Comment (dfuch) [inline test/jdk/java/net/httpclient/Response1xxTest.java:76]: This will not work if the address is the IPv6 loopback. I suggest using the test URIBuilder to build the URI.

**Comment by reschke** [comment]
FWIW, it would be awesome if an equivalent fix could be applied to HTTPURLConnection.

File: test/jdk/java/net/httpclient/Response1xxTest.java (hunk 6 above)
Comment (jaikiran) [reply test/jdk/java/net/httpclient/Response1xxTest.java]: You are right - I hadn't taken that aspect into account. I have now updated the PR to use the `URIBuilder` test library.

**Comment by jaikiran** [comment]
> FWIW, it would be awesome if an equivalent fix could be applied to HTTPURLConnection.

A brief look at the code in that area suggests that it might be relatively straightforward to fix that too. I'll check it out in more detail.

**Review by dfuch — APPROVED** [review]

**Review by ChrisHegarty — APPROVED** [review]
LGTM.

**Review by Michael-Mc-Mahon — APPROVED** [review]
Looks good.

**Comment by jaikiran** [comment]
In https://github.com/openjdk/jdk/pull/10229 Daniel raised a valid question about how we deal with a 101 response when the request didn't ask for an Upgrade. I've updated this PR to ignore such a response too (as allowed by the spec) and updated the test to include this scenario. Test fails without the change and passes with the change.

**Comment by reschke** [comment]
A 101 response implies that the wire protocol is switching to the protocol specified in the "Upgrade" header field. I don't think it would be wise to ignore out, unless the protocol actually stays the same.

Note that this is indeed an edge case because it implies that the server is misbehaving. Thus, I'd rather raise an exception instead of making assumptions that will likely be incorrect.

**Comment by dfuch** [comment]
> A 101 response implies that the wire protocol is switching to the protocol specified in the "Upgrade" header field. I don't think it would be wise to ignore that, unless the protocol actually stays the same.
> 
> Note that this is indeed an edge case because it implies that the server is misbehaving. Thus, I'd rather raise an exception instead of making assumptions that will likely be incorrect.

I agree with Julian - if we didn't ask for an upgrade we should probably consider that as a protocol violation and close the connection.

**Comment by jaikiran** [comment]
Thank you Julian and Daniel for that input. I've now updated this PR to take into account that receiving an unexpected 101 should be considered a protocol error and subsequently necessary action be taken in the client.

New tests have been added to test this scenario.

File: src/java.net.http/share/classes/jdk/internal/net/http/Stream.java [hunk 7]
```diff
@@ -1238,6 +1238,16 @@ void cancel(IOException cause) {
         cancelImpl(cause);
     }
 
+    @Override
+    void onProtocolError(final String errorMessage) {
```
Comment (dfuch) [inline src/java.net.http/share/classes/jdk/internal/net/http/Stream.java:1242]: I believe this should take an `IOException` as parameter and call `cancelImpl(IOException)`.
It should take care of resetting the stream if needed, and properly account for the ref counting of inflight operation.

File: src/java.net.http/share/classes/jdk/internal/net/http/Http1Exchange.java [hunk 8]
```diff
@@ -502,6 +502,16 @@ void cancel(IOException cause) {
         cancelImpl(cause);
     }
 
+    @Override
+    void onProtocolError(final String errorMessage) {
```
Comment (dfuch) [inline src/java.net.http/share/classes/jdk/internal/net/http/Http1Exchange.java:506]: Should take the IOException/ProtocolException as parameter and call cancelImpl(IOException). cancelImpl should do the right thing.

File: src/java.net.http/share/classes/jdk/internal/net/http/Exchange.java [hunk 9]
```diff
@@ -506,7 +559,21 @@ private CompletableFuture<Response> wrapForUpgrade(CompletableFuture<Response> c
...
+        // not expecting an upgrade, but if the server sends a 101 response then we fail the
+        // request and also let the ExchangeImpl deal with it as a protocol error
+        return cf.thenCompose(r -> {
+            if (r.statusCode == 101) {
+                final String errMsg = "Unexpected 101 response, when not upgrading";
+                assert exchImpl != null : "Illegal state - current exchange isn't set";
+                try {
+                    exchImpl.onProtocolError(errMsg);
+                } catch (Throwable ignore){
+                    // ignored
```
[the remark this answers is not in the document]
Comment (dfuch) [reply src/java.net.http/share/classes/jdk/internal/net/http/Exchange.java:571]: I agree. But I believe you should create the exception outside of the try and pass it both to the exchangeImpl and the returned failed future. The concrete implementation of `onProtocolError` should end up calling `cancelImpl` on the concrete subclass in both cases. Calling `cancelImpl` should take care of proper exception recording and take care of operation ref counting too.

File: test/jdk/java/net/httpclient/Response1xxTest.java [hunk 10]
```diff
@@ -64,9 +68,14 @@ public class Response1xxTest {
     private String http1RequestURIBase;
 
 
-    private HttpServerAdapters.HttpTestServer http2Server;
+    private HttpServerAdapters.HttpTestServer http2Server; // h2c
```
Comment (dfuch) [inline test/jdk/java/net/httpclient/Response1xxTest.java:71]: can we remove the outer class from the declaration?

File: test/jdk/java/net/httpclient/Response1xxTest.java [hunk 11]
```diff
@@ -64,9 +68,14 @@ public class Response1xxTest {
...
     private String http2RequestURIBase;
 
+
+    private SSLContext sslContext;
+    private HttpServerAdapters.HttpTestServer https2Server;  // h2
```
Comment (dfuch) [inline test/jdk/java/net/httpclient/Response1xxTest.java:76]: same here?

File: test/jdk/java/net/httpclient/Response1xxTest.java [hunk 12]
```diff
@@ -102,6 +127,10 @@ public void teardown() throws Exception {
             http2Server.stop();
             System.out.println("Stopped HTTP2 server");
         }
+        if (https2Server != null) {
+            https2Server.stop();
+            System.out.println("Stopped (https) HTTP2 server");
+        }
```
Comment (dfuch) [inline test/jdk/java/net/httpclient/Response1xxTest.java:133]: This test should probably use a `final ReferenceTracker TRACKER = ReferenceTracker.INSTANCE;` now to double check that all clients have properly terminated without any dangling operations at the end of the test.

You can `grep TRACKER` in existing HttpClient tests to see how this is used.
The TRACKER can be used to track created clients, and invoked at the end of the test to verify that all resources have been properly reclaimed.

**Review by dfuch — CHANGES_REQUESTED** [review]

**Review by dfuch — APPROVED** [review]
LGTM

**Comment by jaikiran** [comment]
Thank you everyone for the inputs and reviews.

**Comment by reschke** [comment]
Tested in https://github.com/greenbytes/java-http-1xx-tests and found to be working as advertised. Thanks!

**Comment by reschke** [comment]
> We probably should consider how a future API change would work, which allows an application to receive the intermediate responses and make sure this change doesn't cause any problems for that. Though, I don't see an issue currently. Regarding a new API, maybe a method could be added to HttpResponse.BodyHandler to supply intermediate responses?

It would be great if a new ticket would be opened for tracking this enhancement request.

**Comment by jaikiran** [comment]
> Tested in https://github.com/greenbytes/java-http-1xx-tests and found to be working as advertised. Thanks!

Thank you Julian for running those tests.

> > We probably should consider how a future API change would work, which allows an application to receive the intermediate responses and make sure this change doesn't cause any problems for that. Though, I don't see an issue currently. Regarding a new API, maybe a method could be added to HttpResponse.BodyHandler to supply intermediate responses?
> 
> It would be great if a new ticket would be opened for tracking this enhancement request.

That is being tracked in https://bugs.openjdk.org/browse/JDK-8294196
